alchemist-incarnation-collektive/src/main/kotlin/it/unibo/collektive/alchemist/device/CollektiveDevice.kt

```diff
@@ -18,48 +25,69 @@ import it.unibo.collektive.networking.OutboundMessage
  * are retained.
  */
 class CollektiveDevice<P>(
-    private val environment: Environment<Any, P>,
-    override val node: Node<Any>,
+    private val environment: Environment<Any?, P>,
+    override val node: Node<Any?>,
     private val retainMessagesFor: Time,
-) : NodeProperty<Any>, Network, DistanceSensor where P : Position<P> {
-
+) : NodeProperty<Any?>, Network, DistanceSensor where P : Position<P> {
     private data class TimedMessage(val receivedAt: Time, val payload: InboundMessage)
 
     /**
      * The current time.
      */
     var currentTime: Time = Time.ZERO
 
-    private var validMessages: Iterable<TimedMessage> = emptySet()
+    /**
+     * The ID of the node.
+     */
+    val id: ID = IntId(node.id)
 
-    private fun receiveMessage(time: Time, message: InboundMessage) {
+    private val validMessages: MutableList<TimedMessage> = mutableListOf()
+
+    private fun receiveMessage(
+        time: Time,
+        message: InboundMessage,
+    ) {
         validMessages += TimedMessage(time, message)
     }
 
-    override fun cloneOnNewNode(node: Node<Any>) = TODO("Not yet implemented")
+    override fun Aggregate.distances(): Field<Double> =
+        environment.getPosition(node).let { nodePosition ->
+            neighboring(nodePosition).map { position -> nodePosition.distanceTo(position) }
+        }
+
+    override fun cloneOnNewNode(node: Node<Any?>): NodeProperty<Any?> =
+        CollektiveDevice(environment, node, retainMessagesFor)
 
     override fun read(): Set<InboundMessage> {
-        return validMessages
-            .filter { it.receivedAt + retainMessagesFor >= currentTime }
-            .also { validMessages = it }
-            .map { it.payload }
-            .toSet()
+        validMessages.retainAll { it.receivedAt + retainMessagesFor >= currentTime }
+        return validMessages.mapTo(mutableSetOf()) { it.payload }
```

Comment:
```suggestion
        return validMessages.map { it.payload }.toSet()
```

Reply:
nope. That was made for performance optimization.